dxp/liferay_global_event_handler.md

```diff
@@ -0,0 +1,213 @@
+# Liferay’s Global Event Handler
+
+Liferay’s event handler helps the developers for creating/listening/firing/removing custom events that can be listened in whole DXP which is rely on the YUI3 event target implementation. See the implementation [here](https://github.com/liferay/yui3/blob/master/src/event-custom/js/event-target.js#L120).
+
+### First Considerations
+
+First, just consider using the global event handler if necessary. Try to use the default event listener/[callback pattern(if React)](https://reactjs.org/docs/lifting-state-up.html#lifting-state-up) when working. Just use the global event handler when trying to communicate between widgets or not accessible contexts(e.g iframes).
+
+### List of Available Functions
+
+Here is a list of available functions from Liferay’s global variable:
+
+`Liferay.on`
+`Liferay.once`
+`Liferay.onceAfter`
+`Liferay.after`
+`Liferay.before`
+`Liferay.detach`
+`Liferay.detachAll`
+`Liferay.publish`
+`Liferay.fire`
+
+### `Liferay.on`
+
+Description from [yui3/event-target.js at master · liferay/yui3 · GitHub](https://github.com/liferay/yui3/blob/master/src/event-custom/js/event-target.js#L217):
+
+This function subscribes a callback function to a custom event fired by this object or from an object that bubbles its events to this object.
+
+Callback functions for events published with `emitFacade = true` will receive an `EventFacade` as the first argument (typically named “e”). These callbacks can then call `e.preventDefault()` to disable the behaviour published to that event’s `defaultFn` . See the `EventFacade` API for all available properties and methods. Subscribers to non- `emitFacade` events will receive the arguments passed to `fire()` after the event name.
+
+To subscribe to multiple events at once, pass an object as the first argument, where the key:value pairs correspond to the eventName:callback, or pass an array of event names as the first argument to subscribe to all listed events with the same callback.
+
+Returning `false` from a callback is supported as an alternative to calling `e.preventDefault(); e.stopPropagation();` . However, it is recommended to use the event methods whenever possible.
+
+##### Code example:
+
+```js
+var onDestroyPortlet = function () {
+	Liferay.detach('messagePosted', onMessagePosted);
+	Liferay.detach('destroyPortlet', onDestroyPortlet);
+};
+
+Liferay.on('destroyPortlet', onDestroyPortlet);
```

Comment:
I think you should use  a function declaration here, that will let you remove that `var`

Reply:
Given that you're not defining `onMessagePosted` here, it seems extraneous and distracting; I'd just remove it.

I'm also not sure I like the synthetic example here of a `destroyPortlet` listener that does nothing but detach itself (you would have used `once` if that's what what you really wanted to do).

An example like this one seems a bit more appropriate:

```js
Liferay.on('destroyPortlet', (event) => {
    // Perform clean-up here, access event.portletId etc...
});
```

Reply:
Sorry, not a good example that one because I'm omitting the possibility of clean-up.

```js
function destroyPortlet(event) {
    // Perform clean-up here, access event.portletId etc...
}

Liferay.on('destroyPortlet', destroyPortlet);
```

Reply:
> I think you should use a function declaration here, that will let you remove that var

I just copied and paste a JSP usage. Would be good point the file, isn't? I think it will look more practical :)

Reply:
I would simplify this one, it doesn't even have to be real code:

```javascript
function doSomething(event) {
    // Do something here when 'something' happens 
}

Liferay.on('something', doSomething);
```

Then show the `.detach` method down there so it does not get mixed up with the `.on` explanation.